File: src/main/java/preponderous/viron/services/LocationService.java
// Copyright (c) 2024 Preponderous Software
// MIT License

package preponderous.viron.services;

import java.util.Arrays;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import preponderous.viron.config.ServiceConfig;
import preponderous.viron.models.Location;

@Service
public class LocationService {
    private final RestTemplateBuilder restTemplateBuilder;
    private final ServiceConfig serviceConfig;

    private final String baseUrl;
    
    @Autowired
    public LocationService(RestTemplateBuilder restTemplateBuilder, ServiceConfig serviceConfig) {
        this.restTemplateBuilder = restTemplateBuilder;
        this.serviceConfig = serviceConfig;

        this.baseUrl = this.serviceConfig.getVironHost() + ":" + serviceConfig.getVironPort() + "/location";
    }

    public List<Location> getAllLocations() {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Location[]> response = restTemplate.exchange(baseUrl + "/get-all-locations", HttpMethod.GET, null, Location[].class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting locations");
        }
        return Arrays.asList(response.getBody());
    }

    public Location getLocationById(int id) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Location> response = restTemplate.exchange(baseUrl + "/get-location-by-id/" + id, HttpMethod.GET, null, Location.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting location by id");
        }
        return response.getBody();
    }

    public List<Location> getLocationsInEnvironment(int environmentId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Location[]> response = restTemplate.exchange(baseUrl + "/get-locations-in-environment/" + environmentId, HttpMethod.GET, null, Location[].class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting locations in environment");
        }
        return Arrays.asList(response.getBody());
    }

    public List<Location> getLocationsInGrid(int gridId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Location[]> response = restTemplate.exchange(baseUrl + "/get-locations-in-grid/" + gridId, HttpMethod.GET, null, Location[].class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting locations in grid");
        }
        return Arrays.asList(response.getBody());
    }

    public Location getLocationOfEntity(int entityId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Location> response = restTemplate.exchange(baseUrl + "/get-location-of-entity/" + entityId, HttpMethod.GET, null, Location.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting location of entity");
        }
        return response.getBody();
    }

    public boolean addEntityToLocation(int entityId, int locationId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Boolean> response = restTemplate.exchange(baseUrl + "/add-entity-to-location/" + entityId + "/" + locationId, HttpMethod.POST, null, Boolean.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error adding entity to location");
        }
        return response.getBody();
    }

    public boolean removeEntityFromLocation(int entityId, int locationId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Boolean> response = restTemplate.exchange(baseUrl + "/remove-entity-from-location/" + entityId + "/" + locationId, HttpMethod.POST, null, Boolean.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error removing entity from location");
        }
        return response.getBody();
    }

    public boolean removeEntityFromCurrentLocation(int entityId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Boolean> response = restTemplate.exchange(baseUrl + "/remove-entity-from-current-location/" + entityId, HttpMethod.POST, null, Boolean.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error removing entity from current location");
        }
        return response.getBody();
    }

    // testing
    public static void main(String[] args) {
        ServiceConfig serviceConfig = new ServiceConfig();
        serviceConfig.setVironHost("http://localhost");
        serviceConfig.setVironPort(8080);
        LocationService locationService = new LocationService(new RestTemplateBuilder(), serviceConfig);
        List<Location> locations = locationService.getLocationsInEnvironment(1);
        for (Location location : locations) {
            System.out.println(location);
        }
    }
}

File: src/main/python/preponderous/viron/services/locationService.py
from typing import List, Optional
import logging
import requests
from dataclasses import dataclass
from src.main.python.preponderous.viron.models.location import Location

@dataclass
class ServiceConfig:
    viron_host: str
    viron_port: int

class LocationService:
    def __init__(self, service_config: ServiceConfig):
        self.logger = logging.getLogger(__name__)
        self.base_url = f"{service_config.viron_host}:{service_config.viron_port}/api/v1/locations"

    def get_all_locations(self) -> List[Location]:
        response = requests.get(self.base_url)
        response.raise_for_status()
        return response.json()

    def get_location_by_id(self, id: int) -> Location:
        response = requests.get(f"{self.base_url}/{id}")
        if response.status_code == 404:
            raise Exception(f"Location not found with id: {id}")
        response.raise_for_status()
        return response.json()

    def get_locations_in_environment(self, environment_id: int) -> List[Location]:
        response = requests.get(f"{self.base_url}/environment/{environment_id}")
        response.raise_for_status()
        return response.json()

    def get_locations_in_grid(self, grid_id: int) -> List[Location]:
        response = requests.get(f"{self.base_url}/grid/{grid_id}")
        response.raise_for_status()
        return response.json()

    def get_location_of_entity(self, entity_id: int) -> Location:
        response = requests.get(f"{self.base_url}/entity/{entity_id}")
        if response.status_code == 404:
            raise Exception(f"Location not found for entity: {entity_id}")
        response.raise_for_status()
        return response.json()

    def add_entity_to_location(self, entity_id: int, location_id: int) -> None:
        response = requests.put(f"{self.base_url}/{location_id}/entity/{entity_id}")
        if response.status_code == 404:
            raise Exception("Location or entity not found")
        response.raise_for_status()

    def remove_entity_from_location(self, entity_id: int, location_id: int) -> None:
        response = requests.delete(f"{self.base_url}/{location_id}/entity/{entity_id}")
        if response.status_code == 404:
            raise Exception("Location or entity not found")
        response.raise_for_status()

    def remove_entity_from_current_location(self, entity_id: int) -> None:
        response = requests.delete(f"{self.base_url}/entity/{entity_id}")
        if response.status_code == 404:
            raise Exception("Entity not found")
        response.raise_for_status()